### DISCUSSION TITLE
CS8601 and MemberNotNull with ref parameter

### DISCUSSION THREAD

**Discussion by brinko99** [dotnet/csharplang#7172] [Q&A] [answered]
I have the need to apply `MemberNotNull` to a property setter and see a surprising result. I'm seeing:

"CS8601: Possible null reference assignment" on `ref _name`
"CS8774: Member '_name' must have a non-null value when exiting" on `set`

`Set` is a generic method defined in an external library in a nullable context.

```
private string _name;

[MemberNotNull(nameof(_name))]
set
{
    Set(ref _name, value);
}
```

CS8601 appears to be telling me that `_name` must be demonstrated as not-null going into `Set`. This is a surprise to me as `_name` is already guaranteed to be not-null, no? Verifying that `_name` is not-null before `Set` resolves CS8601. I assume `Set` is seen as a `MaybeNull` method with regard to its first parameter. Isn't passing in a null legal in this context anyway?

My question is, why does `MemberNotNull` produce CS8601 in this case at all?

The CS8774 can be resolved by checking the value of `_name` after the `Set` call as expected.

**Comment by Tinister** [comment]
The compiler doesn't know that `Set` doesn't assign null to `ref _name` before the call to `Set` returns.

If the external library author changes `Set` to add `NotNullIfNotNull` like below then your warnings go away.

    static void Set<T>([NotNullIfNotNull("value")] ref T field, T value) { ... }

**Reply by brinko99** [reply]
@Tinister, indeed, that's what I assumed was going on.

However, assigning a value to `_name` before the call to `Set` clears CS8601. Even though `Set` could still assign `null`.

```
[MemberNotNull(nameof(_name))]
set
{
    _name = string.Empty;
    Set(ref _name, value);
}
```

**Reply by TahirAhmadov** [reply, marked as answer]
@brinko99 that's because the `Set` method takes a generic argument and when you set `_name` to a not-null value (which `string.Empty` is), the generic argument becomes `string`, as opposed `string?`. Therefore, the `MemberNotNull` is satisfied.

**Reply by brinko99** [reply]
I see... very subtle!

I remain surprised that setting `_name` before the call to `Set` also clears CS8774. As `Set` is in a nullable oblivious context, it can set `_name` to `null`'. It seems like a guard here should be required.

Thanks

**Reply by TahirAhmadov** [reply]
That would be potentially a warning inside of `Set` method itself, depending on how it's written.
